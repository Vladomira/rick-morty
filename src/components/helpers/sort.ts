import {
  CharacterFullData,
  SortOrder,
  locationObj,
} from '../../types/CharactersData';

export function sortData(items: CharacterFullData[], order: SortOrder) {
  return [...items].sort((a, b) =>
    order === 'asc'
      ? a.episode?.length - b.episode?.length
      : b.episode?.length - a.episode?.length,
  );
}

export function sortLocations(items: locationObj[], order: SortOrder) {
  return [...items].sort((a, b) =>
    order === 'asc' ? a.count - b.count : b.count - a.count,
  );
}
