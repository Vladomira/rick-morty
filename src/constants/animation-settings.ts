export const subMenuAnimate = {
  enter: {
    opacity: 1,
    rotateX: 0,
    transition: {
      duration: 0.5,
    },
    display: "block",
  },
  exit: {
    opacity: 0,
    rotateX: -15,
    transition: {
      duration: 0.5,
      delay: 0.2,
    },
    transitionEnd: {
      display: "none",
    },
  },
};
export const btnsAnimation = {
  initial: { opacity: 0, x: 30 },
  animate: { opacity: 1, x: 0 },
  exit: { opacity: 0, x: 30 },

  transition: {
    ease: "linear",
    opacity: { duration: 1 },
    x: { duration: 1 },
  },
};
export const btnScrollAnimation = {
  show: { opacity: 1, x: 0 },
  animate: { opacity: 1, x: 0 },
  exit: { opacity: 0, x: 60 },

  transition: {
    ease: "linear",
    opacity: { duration: 0.6 },
    x: { duration: 1.3 },
  },

  whileHover: {
    scale: 1.1,
    transition: { duration: 1, scale: { duration: 1, ease: "linear" } },
  },
};
