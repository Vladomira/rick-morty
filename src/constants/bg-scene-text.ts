import { Color, ShaderMaterial } from "three";

export const gradientShaderMaterial = new ShaderMaterial({
  uniforms: {
    color1: { value: new Color("purple") },
    color2: { value: new Color("black") },
  },
  vertexShader: `
    varying vec2 vUv;
    void main() {
      vUv = uv;
      gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);
    }
  `,
  fragmentShader: `
    uniform vec3 color1;
    uniform vec3 color2;
    varying vec2 vUv;
    void main() {
      gl_FragColor = vec4(mix(color1, color2, vUv.y), 1.0);
    }
  `,
});
