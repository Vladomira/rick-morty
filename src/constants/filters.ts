export type FilterName = "gender" | "status" | "species";
export type GenderValue = "Male" | "Female" | "genderless" | "unknown";
export type StatusValue = "Alive" | "Dead" | "unknown";
export type SpeciesValue = "Human" | "Alien" | "Humanoid" | "Cronenberg";

export type FiltersSet = {
  name: FilterName;
  values: (GenderValue | StatusValue | SpeciesValue)[];
};

export const filtersArray: FiltersSet[] = [
  { name: "gender", values: ["Male", "Female", "genderless", "unknown"] },
  { name: "status", values: ["Alive", "Dead", "unknown"] },
  {
    name: "species",
    values: ["Human", "Alien", "Humanoid", "Cronenberg"],
  },
];
