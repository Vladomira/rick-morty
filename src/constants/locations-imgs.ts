import { Euler, Vector3 } from "three";

function scaleValue1(originalValue: number, newWidth: number) {
  return originalValue * (newWidth / 1920);
}

function adjustPoint(originalPoint: number, newWidth: number) {
  const widthDifference = 1920 - newWidth;
  const adjustment = widthDifference * 0.0005;
  return originalPoint + adjustment;
}
export const locationImagesArr = (windowWidth: number) => {
  return [
    // Left
    {
      position: new Vector3(
        scaleValue1(-2.8, windowWidth),
        0,
        adjustPoint(3.7, windowWidth)
      ),
      rotation: new Euler(0, Math.PI / 2.6, 0),
      url: "purge-planet",
    },
    {
      position: new Vector3(
        scaleValue1(-2.8, windowWidth),
        0,
        adjustPoint(2.8, windowWidth)
      ),
      rotation: new Euler(0, Math.PI / 2.5, 0),
      url: "st.-gloopy-noops-hospital",
    },
    {
      position: new Vector3(
        scaleValue1(-2.8, windowWidth),
        0,
        adjustPoint(1.9, windowWidth)
      ),
      rotation: new Euler(0, Math.PI / 2.5, 0),
      url: "mr.-goldenfold's-dream",
    },
    {
      position: new Vector3(
        scaleValue1(-2.8, windowWidth),
        0,
        adjustPoint(0.98, windowWidth)
      ),
      rotation: new Euler(0, Math.PI / 2.5, 0),
      url: "gromflom-prime",
    },
    // top left
    {
      position: new Vector3(-3, 2, 2.2),
      rotation: new Euler(0, 0.7, 0),
      url: "nuptia-4",
    },
    {
      position: new Vector3(-2, 2, 1.6),
      rotation: new Euler(0, 0.7, 0),
      url: "bepis-9",
    },

    //Right
    {
      position: new Vector3(
        scaleValue1(2.8, windowWidth),
        0,
        adjustPoint(3.7, windowWidth)
      ),
      rotation: new Euler(0, -Math.PI / 2.5, 0),
      url: "immortality-field-resort",
    },
    {
      position: new Vector3(
        scaleValue1(2.8, windowWidth),
        0,
        adjustPoint(2.8, windowWidth)
      ),
      rotation: new Euler(0, -Math.PI / 2.5, 0),
      url: "anatomy-park",
    },
    {
      position: new Vector3(
        scaleValue1(2.8, windowWidth),
        0,
        adjustPoint(1.9, windowWidth)
      ),
      rotation: new Euler(0, -Math.PI / 2.5, 0),
      url: "interdimensional-cable",
    },
    {
      position: new Vector3(
        scaleValue1(2.8, windowWidth),
        0,
        adjustPoint(0.98, windowWidth)
      ),
      rotation: new Euler(0, -Math.PI / 2.5, 0),
      url: "abadango",
    },
    // Top - Right ***
    {
      position: new Vector3(3, 2, 2.2),

      rotation: new Euler(0, -0.7, 0),
      url: "earth-(5-126)",
    },
    {
      position: new Vector3(2, 2, 1.6),
      rotation: new Euler(0, -0.7, 0),
      url: "cronenberg-earth",
    },

    // Back
    {
      position: new Vector3(-2, 0, -0.9),
      rotation: new Euler(0, 0, 0),
      url: "bird-world",
    },
    {
      position: new Vector3(-0.6, 0, -0.6),
      rotation: new Euler(0, 0, 0),
      url: "giant's-town",
    },
    {
      position: new Vector3(0.6, 0, -0.6),
      rotation: new Euler(0, 0, 0),
      url: "citadel-of-ricks",
    },
    {
      position: new Vector3(1.9, 0, -0.9),
      rotation: new Euler(0, 0, 0),
      url: "worldender's-lair",
    },
    // top back
    {
      position: new Vector3(-0.7, 1.9, 1.5),
      rotation: new Euler(0, 0, 0),
      url: "venzenulon-7",
    },
    {
      position: new Vector3(0.75, 1.6, 2),

      rotation: new Euler(0, 0, 0),
      url: "earth-(replacement-dimension)",
    },
    // Front
    {
      position: new Vector3(-0.75, 0, 2),

      rotation: new Euler(0, 0, 0),
      url: "earth-(c-137)",
    },
    {
      position: new Vector3(0.8, 0, 2),
      rotation: new Euler(0, 0, 0),
      url: "post-apocalyptic-earth",
    },
  ];
};
