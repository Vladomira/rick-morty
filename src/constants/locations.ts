export const LILAC__COLOR = "#E0ACF9";
export const LIGHT_LILAC__COLOR = "#ECCFFB";
export const BLACK_COLOR = "#151515";
export const PINK_COLOR = "rgb(255,24,240)";
export const YELLOW_COLOR = "#feff26";
