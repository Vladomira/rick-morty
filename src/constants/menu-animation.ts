export const subMenuAnimate = {
  enter: {
    opacity: 1,
    rotateX: 0,
    transition: {
      duration: 0.5,
    },
    display: "block",
  },
  exit: {
    opacity: 0,
    rotateX: -15,
    transition: {
      duration: 0.5,
      delay: 0.2,
    },
    transitionEnd: {
      display: "none",
    },
  },
}
