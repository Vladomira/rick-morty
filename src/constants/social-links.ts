export const socialLinks = [
  { name: "Github", link: "https://github.com/Vladomira" },
  { name: "Linkedin", link: "https://www.linkedin.com/in/julia-vladomira/" },
  { name: "Medium", link: "https://medium.com/@juliavladomira" },
];
