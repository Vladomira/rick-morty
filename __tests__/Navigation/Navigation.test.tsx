import { cleanup } from "@testing-library/react";
import { Links } from "./constants";
import { setupNav } from "./helpers";

describe("Navigation", () => {
  afterEach(() => {
    cleanup();
  });
  it("renders a heading", () => {
    const { heading } = setupNav();
    expect(heading).toBeInTheDocument();
    expect(heading).toHaveTextContent("Rick and Morty");
  });
  it("a heading should have proper classes", () => {
    const { heading, container } = setupNav();
    expect(heading).toHaveClass("header__title shadow-table");
    expect(container).toMatchSnapshot();
  });
  it("nav has 3 child", () => {
    const { nav, container } = setupNav();
    expect(nav).toBeInTheDocument();
    expect(nav.children).toHaveLength(3);
    expect(container).toMatchSnapshot();
  });
  it("links should have proper text", () => {
    const { nav } = setupNav();
    const homeLink = nav.children[0];
    const locationLink = nav.children[1];
    const searchLink = nav.children[2];
    expect(homeLink).toHaveTextContent(Links.HOME);
    expect(locationLink).toHaveTextContent(Links.LOCATION);
    expect(searchLink).toHaveTextContent(Links.SEARCH);
  });
  it("links should have attrgibute href", () => {
    const { nav, container } = setupNav();
    const homeLink = nav.children[0];
    const locationLink = nav.children[1];
    const searchLink = nav.children[2];
    expect(homeLink).toHaveAttribute("href", "/");

    expect(locationLink).toHaveAttribute(
      "href",
      `/${Links.LOCATION.toLowerCase()}`
    );
    expect(searchLink).toHaveAttribute(
      "href",
      `/${Links.SEARCH.toLowerCase()}`
    );
    expect(container).toMatchSnapshot();
  });
});
