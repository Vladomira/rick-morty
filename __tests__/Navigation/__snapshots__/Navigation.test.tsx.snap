// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation a heading should have proper classes 1`] = `
<div>
  <header
    class="header__section"
  >
    <div>
      <h1
        class="header__title shadow-table"
      >
        Rick and Morty
      </h1>
      <nav
        class="mt-3"
        data-testid="navigation"
      >
        <a
          class="nav  "
          href="/"
        >
          <p
            class="hover:drop-shadow-nav "
          >
            Home
          </p>
        </a>
        <a
          class="nav  "
          href="/locations"
        >
          <p
            class="hover:drop-shadow-nav "
          >
            Locations
          </p>
        </a>
        <a
          class="nav  "
          href="/search"
        >
          <p
            class="hover:drop-shadow-nav "
          >
            Search
          </p>
        </a>
      </nav>
    </div>
  </header>
</div>
`;

exports[`Navigation links should have attrgibute href 1`] = `
<div>
  <header
    class="header__section"
  >
    <div>
      <h1
        class="header__title shadow-table"
      >
        Rick and Morty
      </h1>
      <nav
        class="mt-3"
        data-testid="navigation"
      >
        <a
          class="nav  "
          href="/"
        >
          <p
            class="hover:drop-shadow-nav "
          >
            Home
          </p>
        </a>
        <a
          class="nav  "
          href="/locations"
        >
          <p
            class="hover:drop-shadow-nav "
          >
            Locations
          </p>
        </a>
        <a
          class="nav  "
          href="/search"
        >
          <p
            class="hover:drop-shadow-nav "
          >
            Search
          </p>
        </a>
      </nav>
    </div>
  </header>
</div>
`;

exports[`Navigation nav has 3 child 1`] = `
<div>
  <header
    class="header__section"
  >
    <div>
      <h1
        class="header__title shadow-table"
      >
        Rick and Morty
      </h1>
      <nav
        class="mt-3"
        data-testid="navigation"
      >
        <a
          class="nav  "
          href="/"
        >
          <p
            class="hover:drop-shadow-nav "
          >
            Home
          </p>
        </a>
        <a
          class="nav  "
          href="/locations"
        >
          <p
            class="hover:drop-shadow-nav "
          >
            Locations
          </p>
        </a>
        <a
          class="nav  "
          href="/search"
        >
          <p
            class="hover:drop-shadow-nav "
          >
            Search
          </p>
        </a>
      </nav>
    </div>
  </header>
</div>
`;

exports[`Navigation should highlight the active link 1`] = `
<div>
  <header
    class="header__section"
  >
    <div>
      <h1
        class="header__title shadow-table"
      >
        Rick and Morty
      </h1>
      <nav
        class="mt-3"
        data-testid="navigation"
      >
        <a
          class="nav  "
          href="/"
        >
          <p
            class="hover:drop-shadow-nav "
          >
            Home
          </p>
        </a>
        <a
          class="nav  "
          href="/locations"
        >
          <p
            class="hover:drop-shadow-nav "
          >
            Locations
          </p>
        </a>
        <a
          class="nav  "
          href="/search"
        >
          <p
            class="hover:drop-shadow-nav "
          >
            Search
          </p>
        </a>
      </nav>
    </div>
  </header>
</div>
`;
